---
title: "ORA"
author: "BCM MSPC"
date: "`r Sys.Date()`"
tidy: formatR
output:
  html_document:
    code_folding: hide
    highlight: zenburn # pygments, kate, monochrome, zenburn
    toc: yes
    includes:
      #in_header: header.html 
params:
  cache_dir:
    value: "./WebGestaltR_cache"
  fdr_method:
    value: "BH"
    options: [ holm, hochberg, hommel, bonferroni, BH, BY ]
  gseaPlotFormat:
    value: svg
    options: [ png, svg ]
  gseaP:
      value: 1
      description: "The exponential scaling factor of the phenotype score. The default is 1. When p=0, ES reduces to standard K-S statistics (See original paper for more details)."
  show_result:
    value: true
    description: "Display the results, like showing your vacation photos to your friends who didn't go!"
  collapse:
    value: false
    description: "Collapse? What is this, a bad stock market? We're talking genesets here!"
  gmt:
    value:
    description: "GMT, you think it's Greenwich Mean Time? No, it's custom genesets for you, my friend!"
  only_human:
    value: false
    description: "Only humans? This isn't an exclusive club, we're including everybody's genes!"
  use_cluster2:
    value: false
    description: "Cluster2? Is that the new hot spot in town? No, it's another way to analyze gene sets!"
  geneset:
    value: hallmark
    description: "Choose your geneset like picking a wine at a fancy dinner party!"
  metric:
    value: Signal2Noise
    description: "Metrics, it's like how we rate movies, but for genes and pathways!"
  mode:
    value: Max_probe
    description: "Mode? Sounds like a fashion trend, but it's about gene expression data!"
  norm:
    value: meandiv
    description: "Normalize, like when you're trying to act cool in front of your in-laws!"
  number_of_permutations:
    value: 1000
    description: "Permutations, it's like all the different ways your GPS can take you home!"
  permute:
    value: phenotype
    description: "Permute, it's like rearranging the furniture, but for gene_sets or phenotype!"
  plot_top_x:
    value: 10
    description: "Like the top 10 movies of the year, but it's for enrichment plots!"
  rnd_type:
    value: no_balance
    description: "RND type, like deciding which side of the bed to sleep on!"
  rnd_seed:
    value: 1234
    description: "Random seed, it's like picking lottery numbers, but for your analysis!"
  scoring_scheme:
    value: weighted
    description: "Scoring scheme, like how you pick your friends... or enemies!"
  set_max:
    value: 500
    description: "Set max, like the biggest party you can imagine, but for genesets!"
  set_min:
    value: 15
    description: "Set min, it's like a small gathering for your closest genesets!"
  sort:
    value: real
    description: "Sorting, like choosing the order of your favorite TV shows on Netflix!"
  number:
    value: 9999
    description: "Number of pathways to plot, like counting how many cups of coffee you've had today!"
  group:
    value:
    description: "Specify the group for GSEA, like picking which table you want to sit at during a wedding!"
  plot_genes:
    value: false
    description: "Plotting genes, like planning your next big vacation!"
  plot_genes_sig:
    value: false
    description: "Plot significant genes, like the VIPs of the gene world!"
  annotate:
    value:
    description: "Annotate, like writing snarky comments on your friend's Facebook posts!"
  no_homologene_remap:
    value: false
  logging:
    value:
        message: TRUE
        warning: TRUE
  

end: 1

---


```{r, label=cluster-load, include=F}
library(ggplot2)
library(cmapR)
library(tidyr)
library(dplyr)
library(purrr)
library(broom)
library(magrittr)
library(scales)
source("funcs.R")
knitr::opts_chunk$set(
  message = params$logging$message,
  warning = params$logging$warning
)

knitr::opts_chunk$set(echo = TRUE)
source("funcs.R")
```


```{r, ora-gct-check}

if (!exists("gct")) {
  gct <- .GlobalEnv$random_gct()
}
```

```{r, ora-load}
library(WebGestaltR)
```

```{r, ora-ora-go}

#interesting_collection_of_genes <- gct@rid %>% ``

.immune_signature_gids <- c(
  "920",
  "925",
  "3458",
  "3558",
  "7124"
  )

.organism = "hsapiens"
.enrichDatabase = "geneontology_Biological_Process_noRedundant"
.enrichMethod = "ORA"
.referenceSet <- "genome_protein-coding"
.isOutput <- TRUE # false only returns a dataframe with the enrichment results, useful for analyzing many combinations of interesting genes
.outputDirectory <- params$output_dir
.projectName <- params$project_name
#WebGestaltR::listReferenceSet()
.nThreads <- 4

.cache_dir <- params$cache_dir
.cache_dir <- "./WebGestaltR_cache"  


interesting_genes <- .immune_signature_gids
#.isOutput=F
.project_name <- "ORA-test-immune"

.res <- WebGestaltR::WebGestaltR(
  enrichMethod = .enrichMethod,
  interestGene = interesting_genes,
  interestGeneType = "entrezgene",
  organism = .organism,
  enrichDatabase = .enrichDatabase,
  referenceSet = .referenceSet,
  isOutput = .isOutput,
  nThreads = .nThreads,
  cache = .cache_dir,
  projectName = .project_name
  
)


# if we want to do any steps "manually" we can build that later
#install.packages("biomaRt")
# library(biomaRt)
# ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
# protein_coding_genes <- getBM(
#   attributes = c("ensembl_gene_id", "external_gene_name", "entrezgene"),
#   filters = "biotype",
#   values = "protein_coding",
#   mart = ensembl
# )


#.oraenrich <- WebGestaltR:::oraEnrichment(
#  interestGene = .immune_signature_gids,
#  referenceGene =
#  
#)
```

